src/rust-crypto/OutgoingRequestsManager.ts

```diff
@@ -0,0 +1,112 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { OlmMachine } from "@matrix-org/matrix-sdk-crypto-wasm";
+
+import { OutgoingRequest, OutgoingRequestProcessor } from "./OutgoingRequestProcessor";
+import { Logger } from "../logger";
+
+/**
+ * OutgoingRequestsManager: responsible for processing outgoing requests from the OlmMachine.
+ * Ensure that only one loop is going on at once, and that the requests are processed in order.
+ */
+export class OutgoingRequestsManager {
+    /** whether {@link stop} has been called */
+    private stopped = false;
+
+    /** whether a task is currently running */
+    private isTaskRunning = false;
+
+    /** queue of requests to be processed once the current loop is finished */
+    private requestQueue: (() => void)[] = [];
+
+    public constructor(
+        private readonly logger: Logger,
+        private readonly olmMachine: OlmMachine,
+        public readonly outgoingRequestProcessor: OutgoingRequestProcessor,
+    ) {}
+
+    /**
+     * Shut down as soon as possible the current loop of outgoing requests processing.
+     */
+    public stop(): void {
+        this.stopped = true;
+    }
+
+    /**
+     * Process the outgoing requests from the OlmMachine.
+     *
+     * This should be called at the end of each sync, to process any requests that have been queued.
+     * In some cases if outgoing requests need to be sent immediately, this can be called directly.
+     *
+     * There is only one request running at once, and the others are queued.
+     * If a request is currently running the queued request will only trigger an additional run.
+     */
+    public async doProcessOutgoingRequests(): Promise<void> {
+        if (this.isTaskRunning) {
+            // If the task is running, add the request to the queue and wait for completion.
+            // This ensures that the requests are processed only once at a time.
+            await new Promise<void>((resolve) => {
+                this.requestQueue.push(resolve);
+            });
+        } else {
+            await this.executeTask();
+        }
+    }
+
+    private async executeTask(): Promise<void> {
+        this.isTaskRunning = true;
+
+        try {
+            await this.processOutgoingRequests();
+        } finally {
+            this.isTaskRunning = false;
+        }
+
+        if (this.requestQueue.length > 0) {
+            if (this.stopped) {
+                this.requestQueue.forEach((resolve) => resolve());
+                return;
+            }
+            // there are a pending request that need to be executed
+            const awaitingRequests = this.requestQueue.map((resolve) => resolve);
```

Comment:
is this call to `map` just copying the array? It's not an idiom I'm familiar with: at first glance it looks like it might be trying to run the `resolve` callbacks. (Also, I suspect it's not that efficient.)

I don't think there's actually any need to copy it, since we're about to replace it with a new array anyway. But if you *did* need to copy it, I would use a spread operator and write a comment:

```suggestion
            // take a copy of the queue before we reset it.
            const awaitingRequests = [...this.requestQueue];
```

Reply:
removed in favor of deferred